ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/committing/ProcessTransactionSubmission.scala

```diff
@@ -271,15 +271,29 @@ private[kvutils] class ProcessTransactionSubmission(
         }
         ._1
 
-      r <- if (allUnique)
+      _ <- if (allUnique)
         pass
       else
         reject(
           recordTime,
           buildRejectionLogEntry(
             transactionEntry,
             RejectionReason.Disputed("DuplicateKey: Contract Key not unique")))
-    } yield r
+
+      // LookupByKey nodes themselves don't actually fetch the contract.
+      // Therefore we need to do an additional check on all contract keys
+      // that the referred contract satisfies the causal monotonicity invariant.
+      causalKeyMonotonicity = startingKeys.forall { key =>
+        val state = damlState(key)
+        val keyActiveAt =
+          Conversions.parseTimestamp(state.getContractKeyState.getActiveAt).toInstant
+        !keyActiveAt.isAfter(transactionEntry.ledgerEffectiveTime.toInstant)
+      }
```

Comment:
Could you shortly elaborate (in the comment perhaps) on the causal monotonicity definition? I read "all keys of contracts that are inputs of a transaction must NOT be active after the ledger effective time of that transaction". Is that correct and why is that (e.g. I'm thinking of non-consuming)?

Reply:
Also, I agree about correctness-first but I think it's also important to design with performance in mind. What's your take on that, i.e. how would you model the additional load required by the check and how did you plan for performance to be improved later (if at all needed / possible)?

Reply:
The `active_at` field depicts the `ledger_effective_time` of the transaction that created the contract with that particular key. The causal monotonicity requirement defines that all contracts that are an input to a transaction must be active at a ledger effective time <= the ledger effective time as the transaction under validation.

Reply:
> The `active_at` field depicts the `ledger_effective_time` of the transaction that created the contract with that particular key. The causal monotonicity requirement defines that all contracts that are an input to a transaction must be active at a ledger effective time <= the ledger effective time as the transaction under validation.

OK, why not calling it `ledger_effective_time`?

Reply:
This is in line with https://github.com/digital-asset/daml/blob/8e94c7af251eee175363b8b7f45e481a0aca90f7/ledger/participant-state/kvutils/src/main/protobuf/daml_kvutils.proto#L388-L390

Reply:
For clarity I'd still note that it's the `ledger_effective_time` of the transaction that created the contract, at least in the comment of both protos (but agreed, maybe not in this PR).

Reply:
I added comments to both fields in 14a595f.